File: A2Z_striver/patterns/pattern_10/pattern_10.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print a Half Diamond Pattern (Single Peak).
// Example Input: n = 3
// Example Output:
// *
// **
// ***
// **
// *
// Goal: O(N) Rows | O(N^2) Time

class Solution
{
public:
    // ~ Approach 1: Split Loops (Your Approach)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Readable, easier to reason about bounds.
    void pattern10_split(int n)
    {
        // 1. Growing Phase (1 to n)
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j <= i; j++)
            {
                cout << "*";
            }
            cout << "\n";
        }

        // 2. Shrinking Phase (n-1 down to 1)
        // Note: Logic (n-i-1) correctly handles the decrease
        for (int i = 0; i < n - 1; i++)
        {
            for (int j = 0; j < n - i - 1; j++)
            {
                cout << "*";
            }
            cout << "\n";
        }
    }

    // ~ Approach 2: Single Loop (Mathematical)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Concise, demonstrates index manipulation skills.
    void pattern10_single_loop(int n)
    {
        // ! Run for 2*n - 1 rows
        for (int i = 1; i <= 2 * n - 1; i++)
        {

            // ! Logic:
            // If i <= n: stars = i
            // If i > n: stars = 2*n - i
            int stars = i;
            if (i > n)
            {
                stars = 2 * n - i;
            }

            for (int j = 0; j < stars; j++)
            {
                cout << "*";
            }
            cout << "\n";
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 3;

    cout << "--- Approach 1: Split Loops ---\n";
    sol.pattern10_split(n);

    cout << "\n--- Approach 2: Single Loop ---\n";
    sol.pattern10_single_loop(n);

    return 0;
}

File: A2Z_striver/1 Basics/1.2 Patterns/pattern_10/pattern_10.py
import sys

class Solution:
    """
    Problem: Print a Half Diamond Pattern (Single Peak).
    Example Input: n = 3
    Example Output:
    *
    **
    ***
    **
    *
    Goal: Print the pattern using different algorithmic strategies.
    """

    # // ~ Approach 1: Split Loops (Intuitive/Brute Force)
    # // Analysis: Time O(N^2) | Space O(1)
    def pattern10_split(self, n):
        # // - Part 1: Growing Phase (1 to n)
        for i in range(1, n + 1):
            for j in range(i):
                print("*", end="")
            print()
            
        # // - Part 2: Shrinking Phase (n-1 down to 1)
        for i in range(n - 1, 0, -1):
            for j in range(i):
                print("*", end="")
            print()

    # // ~ Approach 2: Single Loop (Mathematical/Optimized)
    # // Analysis: Time O(N^2) | Space O(1)
    def pattern10_single_loop(self, n):
        # // ! Critical: Run for 2*n - 1 rows
        for i in range(2 * n - 1):
            
            # // - Logic: Calculate stars based on 'i'
            # //   If i < n: stars = i + 1
            # //   If i >= n: stars = 2*n - i - 1
            stars = i + 1
            if i >= n:
                stars = 2 * n - i - 1
            
            for j in range(stars):
                print("*", end="")
            print()

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    n = 3
    
    print(f"--- Approach 1: Split Loops (n={n}) ---")
    solver.pattern10_split(n)
    
    print(f"\n--- Approach 2: Single Loop (n={n}) ---")
    solver.pattern10_single_loop(n)